Add ecovacs component

OverloadUT opened the pull request:
## Description:
Adds support for Ecovacs vacuums (DEEBOT vacuums) via an `ecovacs` component that creates `vacuum` entities for each vacuum in the user's account.

Supports pretty much the full vacuum feature-set, and the vacuums respond _instantly_ to Hass service calls due to the always-open XMPP connection, which is pretty cool! It also supports showing the lifespan of the various replaceable components as attributes on the vacuum state.

Support for the reverse-engineered API is very early, and as such it's likely that not all vacuums and not all regions will work correctly on this first release. The Hass documentation will link to the `sucks` project for users to find more information about what vacuums and regions have been tested.

This PR has been tested with two DEEBOT M80 Pros on the same account.

**Related issue (if applicable):** n/a

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.io#5922

## Example entry for `configuration.yaml` (if applicable):
```yaml
ecovacs:
  username: !secret ecovacs_username
  password: !secret ecovacs_password
  country: us
  continent: na
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

OverloadUT committed:
Ecovacs Deebot vacuums WIP

OverloadUT committed:
More WIP

OverloadUT committed:
All core features implemented

Getting fan speed and locating the vac are still unsupported until sucks adds support

OverloadUT committed:
Move init queries to the added_to_hass method

OverloadUT committed:
Adding support for subscribing to events from the sucks library

This support does not exist in sucks yet; this commit serves as a sort of TDD approach of what such support COULD look like.

OverloadUT committed:
Add OverloadUT as ecovacs code owner

OverloadUT committed:
Full support for Ecovacs vacuums (Deebot)

OverloadUT committed:
Add requirements

OverloadUT committed:
Linting fixes

OverloadUT committed:
Make API Device ID random on each boot

OverloadUT committed:
Fix unique ID

Never worked before, as it should have been looking for a key, not an attribute

OverloadUT committed:
Fix random string generation to work in Python 3.5 (thanks, Travis!)

OverloadUT committed:
Add new files to .coveragerc

balloob commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD, \
+    EVENT_HOMEASSISTANT_STOP
+
+REQUIREMENTS = ['sucks==0.9.0']
```
🤣

balloob commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+
+def setup(hass, config):
+    """Set up the Ecovacs component."""
+    _LOGGER.info("Creating new Ecovacs component")
```
debug

balloob commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+    """Set up the Ecovacs component."""
+    _LOGGER.info("Creating new Ecovacs component")
+
+    if ECOVACS_DEVICES not in hass.data:
```
Components are only setup once, no need for this if-statement.

balloob commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+
+    if hass.data[ECOVACS_DEVICES]:
+        _LOGGER.debug("Starting vacuum components")
+        # discovery.load_platform(hass, "sensor", DOMAIN, {}, config)
```
Comment?

balloob commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+CONF_COUNTRY = "country"
+CONF_CONTINENT = "continent"
+
+CONFIG_SCHEMA = vol.Schema({
```
Any reason you went for configuration via configuration.yaml and not using our shiny new config entry?

OverloadUT replied:
You know what, I created the bulk of this component 9 months ago and when I came back to it I didn't think to convert it. I'll read up on the config entry docs and see about converting it

balloob replied:
Hit me up on Discord, would love to help you with this. Hue is a good (although maybe a bit complicated) example.

balloob commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,222 @@
+
+ECOVACS_FAN_SPEED_LIST = ['normal', 'high']
+
+# These consts represent bot statuses that can come from the `sucks` library
```
Would we be able to get them from the sucks lib directly instead of defining them? `sucks.STATUS_AUTO` etc?

OverloadUT replied:
In doing this, I actually decided to move the is_cleaning and is_charging logic in to the sucks lib which removed the need to define or import these at all!

balloob commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,222 @@
+STATUS_ERROR = 'error'
+
+# Any status that represents active cleaning
+STATUSES_CLEANING = [STATUS_AUTO, STATUS_EDGE, STATUS_SPOT, STATUS_SINGLE_ROOM]
```
Make it a set for faster lookups 👍

balloob commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,222 @@
+        _LOGGER.debug("Vacuum initialized: %s", self.name)
+
+    async def async_added_to_hass(self) -> None:
+        """d."""
```
?

balloob commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,222 @@
+    async def async_added_to_hass(self) -> None:
+        """d."""
+        # Fire off some queries to get initial state
+        self.device.statusEvents.subscribe(self.on_status)
```
Since all you do is call `schedule_update_ha_state`, why not use lambdas?

```python
self.device.statusEvents.subscribe(lambda _: self.schedule_update_ha_state())
```

OverloadUT replied:
Yep, that's better. Each of those event handlers used to do more, but I moved the logic to the `sucks` library since their creation. I'll make this change.

balloob commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,222 @@
+        self.device.run(Charge())
+
+    @property
+    def battery_icon(self):
```
This seems a flaw in the vacuum ABC if you need to implement this yourself. Your implementation seems like how it should be?

OverloadUT replied:
It definitely is an issue with the ABC having no formal support for determining if the bot is charging. The reason I chose not to fix the ABC here is because the proper fix is the rearchitecture discussed here: https://github.com/home-assistant/architecture/issues/29

I plan on updating this component after/alongside @cnrd makes the foundational changes

MartinHjelmare commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+    # Convenient hack for debugging to pipe sucks logging to the Hass logger
+    if _LOGGER.getEffectiveLevel() <= logging.DEBUG:
+        import sucks
+        sucks.logging = _LOGGER
```
It would be good if the sucks library could be updated to use a logger per module, as is recommended.

https://docs.python.org/3/library/logging.html#logger-objects

OverloadUT replied:
Yep, you're right. Done.

MartinHjelmare commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,95 @@
+    ecovacs_api = EcoVacsAPI(ECOVACS_API_DEVICEID,
+                             config[DOMAIN].get(CONF_USERNAME),
+                             EcoVacsAPI.md5(config[DOMAIN].get(CONF_PASSWORD)),
+                             config[DOMAIN].get(CONF_COUNTRY).lower(),
```
Move the lowercase modification to the config schema and use `vol.Lower()`.

OverloadUT committed:
Code review changes

(Will require a sucks version bump in a coming commit; waiting for it to release)

OverloadUT commented:
Okay, all review comments addressed except for using Config Entry, which I will work with Balloob on implementing.

rytilahti commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,195 @@
+    def __init__(self, device):
+        """Initialize the Ecovacs Vacuum."""
+        self.device = device
+        self.device.connect_and_wait_until_ready()
```
I/O should be avoided in `__init__` as it blocks the event loop, better connect inside update or before adding the device.

MartinHjelmare replied:
In this case it doesn't block the event loop since we instantiate the entity in `setup_platform` which is run in the thread pool.

But I like to avoid I/O in init on a general bases anyway.

OverloadUT replied:
Would the preferred implementation to be to connect to the device in `setup_platform` and then pass the already-connected `device` object in to the `__init__`? That's an easy change; just want to confirm that's the best practice here.

trisk replied:
If `connect_and_wait_until_ready()` only depends on the API being available, moving it to `setup_platform` is fine.If the call can fail if vacuum is not available when HA starts, it would be better to allow it to be brought on-line asynchronously in `EcovacsVacuum.update`. You can use `add_devices(..., update_before_add=True)` if important attributes like the name are not available until connected.

andlo commented:
I cant get tihis to work on new Hassio / HassOS installation. Get lots of sleekxmpp errors in the log....

OverloadUT commented:
@andlo In the interest of not cluttering up the code review chatter here, let's take this to another thread. Could you either open a GHI where we can troubleshoot (include said logs, scrubbed of any tokens) or you can hit me up on Discord directly at OverloadUT#0001 so we can troubleshoot

MartinHjelmare commented on homeassistant/components/ecovacs.py:
```diff
@@ -0,0 +1,87 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_COUNTRY): vol.All(vol.Lower, cv.string),
```
Use `vol.Lower()`. Probably have to disable pylint warning.

OverloadUT replied:
Changing `vol.Lower` to `vol.Lower()` doesn't work due to the missing parameter. The way I have it here works and seems to be the way it's used elsewhere. Can you clarify what change you're proposing here?

MartinHjelmare replied:
Does it actually not work or is it just pylint complaining. Last time I read the voluptuous docs it's `vol.Lower()` that is correct.

OverloadUT replied:
It throws an exception; not just a linting error
```TypeError: Lower() missing 1 required positional argument: 'v'```

frenck commented:
I was unable to spot a related documentation PR, hence, I've added the `docs-missing` label.

OverloadUT committed:
Bump sucks to 0.9.1 now that it has released

OverloadUT committed:
Update requirements_all.txt as well

balloob commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,195 @@
+
+        data[ATTR_ERROR] = self._error
+
+        for key, val in self.device.components.items():
```
It's not clear to me what values are added to the state machine. Meaning it can't be documented nor can it be judged if they are appropriate to be stored as state attributes to begin with.

OverloadUT replied:
Ah, these are attributes representing the lifespan of replaceable components on the vacuum. Today, they are hard coded in the lib to be `main_brush`, `side_brush`, and `filter`. (In Hass the attributes will be `component_main_brush`, etc.) The value is a 0-100 integer representing the percentage of lifespan remaining. I made it dynamic here because those three components may not be present on all vacuum models, so as we get feedback and make `sucks` more comprehensive for all vacuum models, this should appropriately populate based on what the vacuum actually has.

(I plan on documenting all of this in the docs once I implement the config entries)

balloob commented:
Looks good to me. Can be merged when the docs are done.

OverloadUT committed:
Bump sucks version to fix lifespan value errors

OverloadUT commented:
Docs added, but now we're waiting on an upstream version bump due to a bug I found writing the docs :)

cnrd commented:
@OverloadUT The STATE support has been merged here: https://github.com/home-assistant/home-assistant/commit/2ff5b4ce9547e1d0b15904ee1eb74906eb8f7608 but I'm hoping that I can also get the split of start / pause merged soon: https://github.com/home-assistant/home-assistant/pull/15751 not sure if you want to base your PR on those changes?

OverloadUT committed:
Revert to sucks 0.9.1 and include a fix for a bug in that release

Sucks is being slow to release currently, so doing this so we can get a version out the door.

OverloadUT commented:
Okay, I just pushed up a very tiny change that fixes a bug in sucks 0.9.1 and bumps the requirement back down to that version.

I was originally waiting for 0.9.2 to be released, but the lib owner seems to currently be unavailable.

As soon as the documentation PR is reviewed, we should be good here: https://github.com/home-assistant/home-assistant.io/pull/5922

( @cnrd in the interest of getting a version out so that we can start getting feedback on stability, I'm going to request this version be merged, but I'll work on a refactor to use your new system, if that's agreeable to everyone)

OverloadUT committed:
Switch state_attributes to device_state_attributes

OverloadUT commented on homeassistant/components/vacuum/ecovacs.py:
```diff
@@ -0,0 +1,199 @@
+        self.device.run(VacBotCommand(command, params))
+
+    @property
+    def state_attributes(self):
```
Thanks; I hadn't realized about that method!

leofuscaldi commented:
After this, whats is the next step to have it working with HA?

trisk commented:
@leofuscaldi Install HA from the dev branch or wait for the next release (0.77).
